File: ios/Sources/DownloadPlugin.swift
import DownloadManagerKit
import SwiftRs
import Tauri
import WebKit

class PathArgs: Decodable {
   let path: String
}

/// Settings pushed by the Rust plugin at startup.
///
/// Named for the payload rather than the setting so a second builder option is a new
/// field here, not a second command.
///
/// Every field is optional because the command is invoked unconditionally: an absent
/// value means "keep the platform default", not a malformed call.
class ConfigArgs: Decodable {
   let userAgent: String?
   let storeDir: String?
}

class CreateArgs: Decodable {
   let path: String
   let url: String
   let options: CreateOptions
}

class DownloadPlugin: Plugin {
   /// Computed rather than stored, so constructing the plugin does not construct the
   /// manager. `configure` has to set the store directory before anything reaches
   /// `shared`, because the manager's store opens its file in its own initializer.
   ///
   /// Thread-safe lazy construction still comes from `static let shared` rather than
   /// from a `lazy var`, which is not thread-safe on a class.
   var downloadManager: DownloadManager { DownloadManager.shared }

   /// Subscribed here rather than in `init`, mirroring Android's `load(webView)`.
   ///
   /// Moving it off `init` is what keeps the manager unconstructed until `configure`
   /// runs. Nothing is missed by waiting: `trigger` delivers to listeners registered
   /// from JS, so before a webview exists there are none and it is already a no-op.
   override func load(webview: WKWebView) {
      Task {
          for await download in self.downloadManager.changed {
             try? self.trigger("changed", data: download);
#if DEBUG
             Logger.debug("[\(download.path.lastPathComponent)] \(download.status) - \(String(format: "%.0f", download.progress))% (\(download.receivedBytes)/\(download.totalBytes.map { String($0) } ?? "unknown") bytes)")
#endif
          }
      }
   }

   /// Applies the settings from the Rust plugin's builder.
   ///
   /// Invoked by the Rust plugin during setup rather than from the webview: Tauri
   /// fills the config it hands to `load(webview:)` from `tauri.conf.json`, so a
   /// value set on the Rust builder can only arrive as a command.
   ///
   /// This is also where the download manager is first built, which is load-bearing
   /// rather than incidental. The store directory has to be set before `shared` is
   /// touched, and Tauri defers `load(webview:)` until the webview exists — after
   /// plugin registration — so this command runs first. The Rust side invokes it
   /// unconditionally, with both fields nil when nothing is configured, to keep that
   /// true and to leave the background `URLSession` restoring as early as it does now.
   @objc public func configure(_ invoke: Invoke) throws {
      let args = try invoke.parseArgs(ConfigArgs.self)

      // Before the `Task`, and before any use of `downloadManager`: the manager builds
      // its store — reading the file — during its own initialization.
      if let storeDir = args.storeDir {
         DownloadManager.setStoreDirectory(URL(fileURLWithPath: storeDir, isDirectory: true))
      }

      Task {
         await self.downloadManager.setUserAgent(args.userAgent)
         // No response payload anchors this call inside the Task the way every other
         // handler's does, so hoisting it out would still compile — and would let
         // Rust's blocking `run_mobile_plugin` return before the actor write lands.
         invoke.resolve()
      }
   }

   @objc public func list(_ invoke: Invoke) {
      Task {
         let response = await self.downloadManager.list()
         invoke.resolve(response)
      }
   }

   @objc public func get(_ invoke: Invoke) throws {
      let args = try invoke.parseArgs(PathArgs.self)
      let path = try parsePath(args.path)
      Task {
         let response = await self.downloadManager.get(path: path)
         invoke.resolve(response)
      }
   }
   
   @objc public func create(_ invoke: Invoke) throws {
      let args = try invoke.parseArgs(CreateArgs.self)
      let path = try parsePath(args.path)
      let url = try parseURL(args.url)
      Task {
         let response = await self.downloadManager.create(
            path: path,
            url: url,
            options: args.options
         )
         invoke.resolve(response)
      }
   }
   
   @objc public func start(_ invoke: Invoke) throws {
      let args = try invoke.parseArgs(PathArgs.self)
      let path = try parsePath(args.path)
      Task {
         do {
            let response = try await self.downloadManager.start(path: path)
            invoke.resolve(response)
         } catch {
            invoke.reject(error.localizedDescription)
         }
      }
   }
   
   @objc public func cancel(_ invoke: Invoke) throws {
      let args = try invoke.parseArgs(PathArgs.self)
      let path = try parsePath(args.path)
      Task {
         do {
            let response = try await self.downloadManager.cancel(path: path)
            invoke.resolve(response)
         } catch {
            invoke.reject(error.localizedDescription)
         }
      }
   }
   
   @objc public func pause(_ invoke: Invoke) throws {
      let args = try invoke.parseArgs(PathArgs.self)
      let path = try parsePath(args.path)
      Task {
         do {
            let response = try await self.downloadManager.pause(path: path)
            invoke.resolve(response)
         } catch {
            invoke.reject(error.localizedDescription)
         }
      }
   }
   
   @objc public func resume(_ invoke: Invoke) throws {
      let args = try invoke.parseArgs(PathArgs.self)
      let path = try parsePath(args.path)
      Task {
         do {
            let response = try await self.downloadManager.resume(path: path)
            invoke.resolve(response)
         } catch {
            invoke.reject(error.localizedDescription)
         }
      }
   }
}

@_cdecl("init_plugin_download")
func initPlugin() -> Plugin {
   return DownloadPlugin()
}

File: android/src/main/java/org/silvermine/plugin/download/DownloadPlugin.kt
package org.silvermine.plugin.download

import android.app.Activity
import android.util.Log
import android.webkit.WebView
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import org.silvermine.downloadmanager.CreateOptions
import org.silvermine.downloadmanager.DownloadManager
import org.silvermine.downloadmanager.parsePath
import org.silvermine.downloadmanager.parseURI
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.json.JSONArray
import java.io.File

@InvokeArg
class PathArgs {
   var path: String? = null
}

/**
 * Bridge-local mirror of [CreateOptions], kept separate so Tauri's `@InvokeArg`
 * reflection and kotlinx.serialization never share a type.
 *
 * Nullable so an absent value is rejected rather than defaulted. The Rust bridge
 * resolves the API default before invoking, so the policy always arrives stated.
 */
@InvokeArg
class CreateOptionsArgs {
   var allowMetered: Boolean? = null
}

/**
 * Settings pushed by the Rust plugin at startup.
 *
 * Named for the payload rather than the setting so a second builder option is a new
 * field here, not a second command.
 */
@InvokeArg
class ConfigArgs {
   var userAgent: String? = null
   var storeDir: String? = null
}

@InvokeArg
class CreateArgs {
   var path: String? = null
   var url: String? = null
   var options: CreateOptionsArgs? = null
}

@TauriPlugin
class DownloadPlugin(activity: Activity) : Plugin(activity) {
   private val json = Json { encodeDefaults = true }

   /**
    * Held because `Plugin.activity` is `private`, so a subclass can reach it from an
    * initializer — where the constructor parameter is still in scope — but not from a
    * member. The application context rather than the activity, which must not outlive
    * its own lifecycle.
    */
   private val appContext = activity.applicationContext

   /**
    * Computed rather than `by lazy`: [configure] has to build the instance itself, to
    * pass the store directory into a constructor that reads it. [DownloadManager
    * .getInstance] is already a double-checked singleton, so it — not a second lazy
    * delegate racing with it — is the one place construction is synchronized.
    */
   private val downloadManager: DownloadManager
      get() = DownloadManager.getInstance(appContext)

   private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

   override fun load(webView: WebView) {
      scope.launch {
         downloadManager.changed.collect { item ->
            try {
               trigger("changed", JSObject(json.encodeToString(item)))
               Log.d(
                  TAG,
                  "[${File(item.path).name}] ${item.status} - ${"%.0f".format(item.progress)}%" +
                     " (${item.receivedBytes}/${item.totalBytes ?: "unknown"} bytes)",
               )
            } catch (e: Exception) {
               Log.e(TAG, "Failed to emit changed event: ${e.message}")
            }
         }
      }
   }

   override fun onDestroy() {
      super.onDestroy()
      scope.cancel()
   }

   /**
    * Applies the settings from the Rust plugin's builder.
    *
    * Invoked by the Rust plugin during setup rather than from the webview: Tauri
    * fills the config it hands to [load] from `tauri.conf.json`, so a value set on
    * the Rust builder can only arrive as a command.
    *
    * This is where the download manager is built, and that is load-bearing rather than
    * incidental. The store directory is read by [DownloadStore]'s constructor, so it
    * has to reach [DownloadManager.getInstance] on the call that creates the singleton.
    * Tauri defers [load] until the webview exists, which is after plugin registration,
    * so this command runs first — and the Rust side invokes it unconditionally, with
    * both fields null when nothing is configured, to keep that true.
    *
    * Every argument is optional: absent means "keep the platform default", not a
    * malformed call.
    */
   @Command
   fun configure(invoke: Invoke) {
      val args = invoke.parseArgs(ConfigArgs::class.java)
      val manager = DownloadManager.getInstance(appContext, args.storeDir?.let { File(it) })

      args.userAgent?.let { manager.userAgent = it }
      invoke.resolve()
   }

   @Command
   fun list(invoke: Invoke) {
      scope.launch {
         try {
            val items = withContext(Dispatchers.IO) { downloadManager.list() }
            val result = JSObject().apply {
               put("value", JSONArray(json.encodeToString(items)))
            }
            invoke.resolve(result)
         } catch (e: Exception) {
            invoke.reject(e.message)
         }
      }
   }

   @Command
   fun get(invoke: Invoke) {
      val args = invoke.parseArgs(PathArgs::class.java)
      val path = try {
         parsePath(args.path ?: throw IllegalArgumentException("Missing required argument: path"))
      } catch (e: Exception) {
         return invoke.reject(e.message)
      }
      scope.launch {
         try {
            val response = withContext(Dispatchers.IO) { downloadManager.get(path) }
            invoke.resolve(JSObject(json.encodeToString(response)))
         } catch (e: Exception) {
            invoke.reject(e.message)
         }
      }
   }

   @Command
   fun create(invoke: Invoke) {
      val args = invoke.parseArgs(CreateArgs::class.java)
      val path = try {
         parsePath(args.path ?: throw IllegalArgumentException("Missing required argument: path"))
      } catch (e: Exception) {
         return invoke.reject(e.message)
      }
      val url = try {
         parseURI(args.url ?: throw IllegalArgumentException("Missing required argument: url"))
      } catch (e: Exception) {
         return invoke.reject(e.message)
      }
      val options = try {
         CreateOptions(
            allowMetered = args.options?.allowMetered
               ?: throw IllegalArgumentException("Missing required argument: options.allowMetered"),
         )
      } catch (e: Exception) {
         return invoke.reject(e.message)
      }

      scope.launch {
         // Guarded like every sibling command. Without this the store's own write
         // failures — an unwritable configured directory reaches `AtomicFile.startWrite`
         // as an IOException — escape to a scope with no handler and take the app down,
         // where desktop returns the error to the caller.
         try {
            val response = withContext(Dispatchers.IO) { downloadManager.create(path, url, options) }
            invoke.resolve(JSObject(json.encodeToString(response)))
         } catch (e: Exception) {
            invoke.reject(e.message)
         }
      }
   }

   @Command
   fun start(invoke: Invoke) {
      val args = invoke.parseArgs(PathArgs::class.java)
      val path = try {
         parsePath(args.path ?: throw IllegalArgumentException("Missing required argument: path"))
      } catch (e: Exception) {
         return invoke.reject(e.message)
      }
      scope.launch {
         try {
            val response = withContext(Dispatchers.IO) { downloadManager.start(path) }
            invoke.resolve(JSObject(json.encodeToString(response)))
         } catch (e: Exception) {
            invoke.reject(e.message)
         }
      }
   }

   @Command
   fun cancel(invoke: Invoke) {
      val args = invoke.parseArgs(PathArgs::class.java)
      val path = try {
         parsePath(args.path ?: throw IllegalArgumentException("Missing required argument: path"))
      } catch (e: Exception) {
         return invoke.reject(e.message)
      }
      scope.launch {
         try {
            val response = withContext(Dispatchers.IO) { downloadManager.cancel(path) }
            invoke.resolve(JSObject(json.encodeToString(response)))
         } catch (e: Exception) {
            invoke.reject(e.message)
         }
      }
   }

   @Command
   fun pause(invoke: Invoke) {
      val args = invoke.parseArgs(PathArgs::class.java)
      val path = try {
         parsePath(args.path ?: throw IllegalArgumentException("Missing required argument: path"))
      } catch (e: Exception) {
         return invoke.reject(e.message)
      }
      scope.launch {
         try {
            val response = withContext(Dispatchers.IO) { downloadManager.pause(path) }
            invoke.resolve(JSObject(json.encodeToString(response)))
         } catch (e: Exception) {
            invoke.reject(e.message)
         }
      }
   }

   @Command
   fun resume(invoke: Invoke) {
      val args = invoke.parseArgs(PathArgs::class.java)
      val path = try {
         parsePath(args.path ?: throw IllegalArgumentException("Missing required argument: path"))
      } catch (e: Exception) {
         return invoke.reject(e.message)
      }
      scope.launch {
         try {
            val response = withContext(Dispatchers.IO) { downloadManager.resume(path) }
            invoke.resolve(JSObject(json.encodeToString(response)))
         } catch (e: Exception) {
            invoke.reject(e.message)
         }
      }
   }

   companion object {
      private const val TAG = "DownloadPlugin"
   }
}
